#3 - Import the required packages with their customary aliases as follows:

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#4 Use the %matplotlib inline magic command so that your plots show in the notebook without having to call plt.show() every time.

In [2]:
%matplotlib inline

#5 Using the pandas read_csv() method, read the GDP dataset into your notebook as a DataFrame called gdp_df. Take a look at the first 6 rows.

In [3]:
gdp_df = pd.read_csv("..\data\gdp_percapita.csv", nrows=6868)

In [4]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6868 entries, 0 to 6867
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6868 non-null   object 
 1   Year             6868 non-null   int64  
 2   Value            6868 non-null   float64
 3   Value Footnotes  1 non-null      float64
dtypes: float64(2), int64(1), object(1)
memory usage: 214.8+ KB


In [5]:
gdp_df.head(6)

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2019  2065.036398              NaN
1     Afghanistan  2018  2033.779002              NaN
2     Afghanistan  2017  2058.383832              NaN
3     Afghanistan  2016  2057.062164              NaN
4     Afghanistan  2015  2068.265904              NaN
5     Afghanistan  2014  2102.385234              NaN

This didn't work: internet_df = pd.read_csv("..\data\internet_use.csv")

Repeat for the internet use dataset. Call this DataFrame internet_df. Take a look at the first six rows.

In [6]:
internet_df = pd.read_csv("../data/internet_use.csv", nrows=4495)

In [7]:
internet_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4495 entries, 0 to 4494
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  4495 non-null   object 
 1   Year             4495 non-null   int64  
 2   Value            4495 non-null   float64
 3   Value Footnotes  948 non-null    float64
dtypes: float64(2), int64(1), object(1)
memory usage: 140.6+ KB


In [8]:
internet_df.head(6)

Country or Area  Year     Value  Value Footnotes
0     Afghanistan  2014  6.390000              NaN
1     Afghanistan  2013  5.900000              NaN
2     Afghanistan  2012  5.454545              NaN
3     Afghanistan  2011  5.000000              NaN
4     Afghanistan  2010  4.000000              NaN
5     Afghanistan  2009  3.550000              NaN

#7 Look at the shape of each DataFrame - how many rows? How many columns?

In [9]:
internet_df.shape

(4495, 4)

In [10]:
gdp_df.shape

(6868, 4)

In [11]:
print(internet_df.shape)

(4495, 4)


#8 Take a look at the datatypes for the columns in each DataFrame.

In [12]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6868 entries, 0 to 6867
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6868 non-null   object 
 1   Year             6868 non-null   int64  
 2   Value            6868 non-null   float64
 3   Value Footnotes  1 non-null      float64
dtypes: float64(2), int64(1), object(1)
memory usage: 214.8+ KB


In [13]:
internet_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4495 entries, 0 to 4494
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  4495 non-null   object 
 1   Year             4495 non-null   int64  
 2   Value            4495 non-null   float64
 3   Value Footnotes  948 non-null    float64
dtypes: float64(2), int64(1), object(1)
memory usage: 140.6+ KB


9. Take a look at the last 10 rows of each DataFrame in turn.

In [14]:
gdp_df.tail(10)

Country or Area  Year        Value  Value Footnotes
6858        Zimbabwe  1999  3600.849291              NaN
6859        Zimbabwe  1998  3653.920016              NaN
6860        Zimbabwe  1997  3580.048793              NaN
6861        Zimbabwe  1996  3520.430146              NaN
6862        Zimbabwe  1995  3226.413930              NaN
6863        Zimbabwe  1994  3263.934978              NaN
6864        Zimbabwe  1993  3033.504852              NaN
6865        Zimbabwe  1992  3054.889178              NaN
6866        Zimbabwe  1991  3426.598094              NaN
6867        Zimbabwe  1990  3324.348171              NaN

In [15]:
internet_df.tail(10)

Country or Area  Year     Value  Value Footnotes
4485        Zimbabwe  2002  3.994356              NaN
4486        Zimbabwe  2001  0.799846              NaN
4487        Zimbabwe  2000  0.401434              NaN
4488        Zimbabwe  1999  0.161676              NaN
4489        Zimbabwe  1998  0.081648              NaN
4490        Zimbabwe  1997  0.033080              NaN
4491        Zimbabwe  1996  0.016790              NaN
4492        Zimbabwe  1995  0.007684              NaN
4493        Zimbabwe  1994  0.001739              NaN
4494        Zimbabwe  1990  0.000000              NaN

10. Drop the value footnotes column from both DataFrames. Check that this worked as expected.

In [16]:
internet_df = internet_df.drop(columns=("Value Footnotes"))
internet_df.head()

Country or Area  Year     Value
0     Afghanistan  2014  6.390000
1     Afghanistan  2013  5.900000
2     Afghanistan  2012  5.454545
3     Afghanistan  2011  5.000000
4     Afghanistan  2010  4.000000

In [17]:
gdp_df = gdp_df.drop(columns=("Value Footnotes")) 

In [18]:
gdp_df.head()

Country or Area  Year        Value
0     Afghanistan  2019  2065.036398
1     Afghanistan  2018  2033.779002
2     Afghanistan  2017  2058.383832
3     Afghanistan  2016  2057.062164
4     Afghanistan  2015  2068.265904

Number 11 - Change the columns for the GDP Per Capita DataFrame to ‘Country’, ‘Year’, and ‘GDP_Per_Capita’.

In [19]:
gdp_df = gdp_df.rename(columns={'Country or Area':'Country', 'Year': 'Year', 'Value': 'GDP_Per_Capita'})

In [20]:
gdp_df.head()

Country  Year  GDP_Per_Capita
0  Afghanistan  2019     2065.036398
1  Afghanistan  2018     2033.779002
2  Afghanistan  2017     2058.383832
3  Afghanistan  2016     2057.062164
4  Afghanistan  2015     2068.265904

Number 12 - Change the columns for the Internet Users DataFrame to ‘Country’, ‘Year’, and ‘Internet_Users_Pct’.

In [21]:
internet_df = internet_df.rename(columns={'Country or Area':'Country', 'Year': 'Year', 'Value': 'Internet_Users_Pct'})
internet_df.head()

Country  Year  Internet_Users_Pct
0  Afghanistan  2014            6.390000
1  Afghanistan  2013            5.900000
2  Afghanistan  2012            5.454545
3  Afghanistan  2011            5.000000
4  Afghanistan  2010            4.000000

13. Merge the two DataFrames to one. Merge all rows from each of the two DataFrames. Call the new DataFrame gdp_and_internet_use
14. Look at the first five rows of your new DataFrame to confirm it merged correctly

In [22]:
gdp_and_internet_use = pd.merge(gdp_df, internet_df, on=['Country', 'Year'], how='outer')
gdp_and_internet_use.head(10)

Country  Year  GDP_Per_Capita  Internet_Users_Pct
0  Afghanistan  2019     2065.036398                 NaN
1  Afghanistan  2018     2033.779002                 NaN
2  Afghanistan  2017     2058.383832                 NaN
3  Afghanistan  2016     2057.062164                 NaN
4  Afghanistan  2015     2068.265904                 NaN
5  Afghanistan  2014     2102.385234            6.390000
6  Afghanistan  2013     2116.465454            5.900000
7  Afghanistan  2012     2075.491747            5.454545
8  Afghanistan  2011     1904.559799            5.000000
9  Afghanistan  2010     1957.029338            4.000000

15. Look at the last five rows to make sure the data is clean and as expected.

In [62]:
gdp_and_internet_use.tail(10)

Country  Year  GDP_Per_Capita  Internet_Users_Pct
7895   Yemen  2004             NaN            0.881223
7896   Yemen  2003             NaN            0.604734
7897   Yemen  2002             NaN            0.518796
7898   Yemen  2001             NaN            0.090802
7899   Yemen  2000             NaN            0.082500
7900   Yemen  1999             NaN            0.056629
7901   Yemen  1998             NaN            0.023323
7902   Yemen  1997             NaN            0.015025
7903   Yemen  1996             NaN            0.000621
7904   Yemen  1990             NaN            0.000000

In [63]:
gdp_and_internet_use.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7905 entries, 0 to 7904
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Country             7905 non-null   object 
 1   Year                7905 non-null   int64  
 2   GDP_Per_Capita      6868 non-null   float64
 3   Internet_Users_Pct  4495 non-null   float64
dtypes: float64(2), int64(1), object(1)
memory usage: 308.8+ KB


16. Subset the combined DataFrame to keep only the data for 2004, 2009, and 2014. Check that this happened correctly.

In [26]:
years_04_09_14= gdp_and_internet_use[gdp_and_internet_use['Year'].isin([2004, 2009, 2014])]
print(years_04_09_14)

                    Country  Year  GDP_Per_Capita  Internet_Users_Pct
5               Afghanistan  2014     2102.385234            6.390000
10              Afghanistan  2009     1758.904043            3.550000
15              Afghanistan  2004     1200.278256            0.105809
23                  Albania  2014    11623.866679           60.100000
28                  Albania  2009    10346.864199           41.200000
...                     ...   ...             ...                 ...
7868  Virgin Islands (U.S.)  2009             NaN           27.396510
7873  Virgin Islands (U.S.)  2004             NaN           27.377009
7885                  Yemen  2014             NaN           22.550000
7890                  Yemen  2009             NaN            9.960000
7895                  Yemen  2004             NaN            0.881223

[830 rows x 4 columns]


17. Create three new DataFrames, one for 2004, one for 2009, and one for 2014. Give them meaningful names that aren't too long.

In [33]:
sub04= years_04_09_14[years_04_09_14['Year']==2004]
sub09= years_04_09_14[years_04_09_14['Year']==2009] 
sub14= years_04_09_14[years_04_09_14['Year']==2014]
print(sub04)
print(sub09)
print(sub14)

                    Country  Year  GDP_Per_Capita  Internet_Users_Pct
15              Afghanistan  2004     1200.278256            0.105809
33                  Albania  2004     7604.838002            2.420388
63                  Algeria  2004    10058.435939            4.634475
93                   Angola  2004     5593.241537            0.464815
123     Antigua and Barbuda  2004    18745.681534           24.266544
...                     ...   ...             ...                 ...
7771     T.F.Y.R. Macedonia  2004             NaN           24.440000
7829              Venezuela  2004             NaN            8.404470
7853               Viet Nam  2004             NaN            7.642409
7873  Virgin Islands (U.S.)  2004             NaN           27.377009
7895                  Yemen  2004             NaN            0.881223

[276 rows x 4 columns]
                    Country  Year  GDP_Per_Capita  Internet_Users_Pct
10              Afghanistan  2009     1758.904043             3.55

18. Which country had the highest percentage of internet users in 2014? What was the percentage? (Try typing the first 3 letters of your DataFrame name and hitting the tab key for auto-complete options).

In [118]:
sub14.sort_values(by=['Internet_Users_Pct'], ascending=False).head(1).filter(items=['Internet_Users_Pct', 'Country'])

Internet_Users_Pct  Country
2625               98.16  Iceland

Answer: Iceland	98.16

19. Which country had the lowest percentage of internet users in 2014? What was the percentage?

In [117]:
sub14.sort_values(by=['Internet_Users_Pct']).head(1).filter(items=['Internet_Users_Pct', 'Country'])

Internet_Users_Pct                      Country
7079                 0.0  Dem. People's Rep. of Korea

Answer: Dem. People's Rep. of Korea	0.0

20. Repeat for 2004 and 2009.

In [116]:
sub04.sort_values(by=['Internet_Users_Pct'], ascending=False).head(1).filter(items=['Internet_Users_Pct', 'Country'])

Internet_Users_Pct Country
6019               83.89  Sweden

Answer: Sweden	83.89

In [115]:
sub04.sort_values(by=['Internet_Users_Pct']).head(1).filter(items=['Internet_Users_Pct', 'Country'])

Internet_Users_Pct                      Country
7088                 0.0  Dem. People's Rep. of Korea

Answer: Dem. People's Rep. of Korea	0.0

In [114]:
sub09.sort_values(by=['Internet_Users_Pct'], ascending=False).head(1).filter(items=['Internet_Users_Pct', 'Country'])

Internet_Users_Pct  Country
2630                93.0  Iceland

Answer: Iceland	93.0

In [113]:
sub09.sort_values(by=['Internet_Users_Pct']).head(1).filter(items=['Internet_Users_Pct', 'Country'])

Internet_Users_Pct                      Country
7083                 0.0  Dem. People's Rep. of Korea

Answer: Dem. People's Rep. of Korea	0.0

21. Which country had the highest gdp per capita in 2014? What was the gdp per capita?

In [122]:
sub14.sort_values(by=['GDP_Per_Capita'], ascending=False).head(1).filter(items=['GDP_Per_Capita', 'Country'])

GDP_Per_Capita           Country
3721    156514.13385  Macao SAR, China

Answer: 156514.13385	Macao SAR, China

22. Which country had the lowest gdp per capita in 2014? What was the gdp per capita?

In [145]:
sub14.sort_values(by=['GDP_Per_Capita'], ascending=False).dropna(subset=['GDP_Per_Capita','Country']).filter(items=['GDP_Per_Capita', 'Country']).tail(1)

GDP_Per_Capita                   Country
1047      822.610458  Central African Republic

Answer: 822.610458  Central African Republic

23. Create some scatterplots:

a. 2004 Percent Using the Internet vs GDP Per Capita

b. 2009 Percent Using the Internet vs GDP Per Capita

c. 2014 Percent Using the Internet vs GDP Per Capita

